File: syntactic-unification/unification.py
# Source from https://blog.witchoflight.com/2020/syntactic-unification/

from dataclasses import dataclass


@dataclass(frozen=True)
class Var:
    name: str

    def __repr__(self) -> str:
        return self.name


def unify(a, b, model={}):
    a = walk(model, a)
    b = walk(model, b)

    if a == b:
        return model

    # ... variable cases ...

    if isinstance(a, Var):
        if occurs(a, b, model):
            return None
        return {**model, a: b}

    if isinstance(b, Var):
        if occurs(a, b, model):
            return None
        return {**model, b: a}

    # ... tuple cases ...

    if isinstance(a, tuple) and isinstance(b, tuple):
        if len(a) != len(b):
            return None

        for ax, bx in zip(a, b):
            model = unify(ax, bx, model)
            if model is None:
                return None
        return model

    return None


def walk(model, term):
    while term in model:
        term = model[term]
    return term


def occurs(a, b, model):
    b = walk(model, b)
    if a == b:
        return True
    if isinstance(b, tuple):
        return any(occurs(a, x, model) for x in b)
    return False

File: syntactic-unification/unification.rb
# frozen_string_literal: true

class Var
  attr_reader :name

  def initialize(name)
    @name = name
  end

  def inspect
    "#{self.class}(name: #{@name.inspect})"
  end
end

def unify(a, b, model = {})
  a = walk(a, model)
  b = walk(b, model)

  return model if a == b

  if a.is_a?(Var)
    return if occurs?(a, b, model)

    return model.merge({ a => b })
  end

  if b.is_a?(Var)
    return if occurs?(a, b, model)

    return model.merge({ b => a })
  end

  if a.is_a?(Array) && b.is_a?(Array)
    return None unless a.length == b.length

    a.zip(b).each do |ax, bx|
      model = unify(ax, bx, model)
      return nil if model.nil?
    end

    return model
  end

  nil
end

def walk(term, model)
  term = model[term] while model.key?(term)
  term
end

def occurs?(a, b, model)
  b = walk(b, model)
  return true if b == a
  return b.any? { |bx| occurs?(a, bx, model) } if b.is_a?(Array)

  false
end
